lib/vtls/nss.c

```diff
@@ -578,17 +578,18 @@ static CURLcode nss_cache_crl(SECItem *crl_der)
   /* acquire lock before call of CERT_CacheCRL() and accessing nss_crl_list */
   PR_Lock(nss_crllock);
 
-  /* store the CRL item so that we can free it in Curl_nss_cleanup() */
-  if(insert_wrapped_ptr(&nss_crl_list, crl_der) != CURLE_OK) {
+  if(SECSuccess != CERT_CacheCRL(db, crl_der)) {
+    /* unable to cache CRL */
     SECITEM_FreeItem(crl_der, PR_TRUE);
     PR_Unlock(nss_crllock);
-    return CURLE_OUT_OF_MEMORY;
+    return CURLE_SSL_CRL_BADFILE;
   }
 
-  if(SECSuccess != CERT_CacheCRL(db, crl_der)) {
-    /* unable to cache CRL */
+  /* store the CRL item so that we can free it in Curl_nss_cleanup() */
+  if(insert_wrapped_ptr(&nss_crl_list, crl_der) != CURLE_OK) {
+    SECITEM_FreeItem(crl_der, PR_TRUE);
```

Comment:
You need to call `CERT_UncacheCRL()` before releasing the memory to avoid a possible double free() later on.  If `CERT_UncacheCRL()` fails, which is unlikely, it is IMO better to leak the memory.  Something like:
```
if (SECSuccess == CERT_UncacheCRL(db, crl_der))
  SECITEM_FreeItem(crl_der, PR_TRUE);
```

Reply:
Ah, nice catch, fixed and rebased.